odxtools/diaglayer.py

```diff
@@ -843,67 +845,90 @@ def _build_coded_prefix_tree(self):
+                            service: DiagService) -> None:
+
+        # make sure that tree has an entry for the given prefix
+        sub_tree = prefix_tree
+        for b in coded_prefix:
+            if b not in sub_tree:
+                sub_tree[b] = {}
+            sub_tree = cast(PrefixTree, sub_tree[b])
+
+        # Store the object as in the prefix tree. This is done by
+        # assigning the list of possible objects to the key -1 of the
+        # dictionary (this is quite hacky...)
+        if sub_tree.get(-1) is None:
+            sub_tree[-1] = [service]
+        else:
+            cast(List[DiagService], sub_tree[-1]).append(service)
+
+    def _find_services_for_uds(self, message: Union[bytes, bytearray]) -> List[DiagService]:
         if not hasattr(self, "_prefix_tree"):
             # Compute the prefix tree the first time this decode function is called.
             self._prefix_tree = self._build_coded_prefix_tree()
         prefix_tree = self._prefix_tree
 
         # Find matching service(s) in prefix tree
-        possible_services = []
+        possible_services: List[DiagService] = []
         for b in message:
-            if prefix_tree.get(b) is not None:
-                assert isinstance(prefix_tree.get(b), dict)
-                prefix_tree = prefix_tree.get(b)
+            if b in prefix_tree:
+                assert isinstance(prefix_tree[b], dict)
+                prefix_tree = cast(PrefixTree, prefix_tree[b])
             else:
                 break
             if -1 in prefix_tree:
-                possible_services += prefix_tree[-1]
+                possible_services += cast(List[DiagService], prefix_tree[-1])
         return possible_services
 
-    def decode(self, message: Union[bytes, bytearray]) -> Iterable[Message]:
+    def decode(self, message: Union[bytes, bytearray]) -> List[Message]:
+        decoded_messages: List[Message] = []
         possible_services = self._find_services_for_uds(message)
-
-        if possible_services is None:
-            raise DecodeError(f"Couldn't find corresponding service for message {message.hex()}.")
-
-        decoded_messages = []
-
         for service in possible_services:
             try:
                 decoded_messages.append(service.decode_message(message))
             except DecodeError as e:
-                pass
+                # check if the message can be decoded as a global
+                # negative response for the service
+                for gnr in self.global_negative_responses:
+                    try:
+                        decoded_gnr = gnr.decode(message)
+                        decoded_messages.append(
+                            Message(
+                                coded_message=message,
+                                service=service,
+                                structure=gnr,
+                                param_dict=decoded_gnr))
+                    except DecodeError:
```

Comment:
I have seen the decode logic throw all kind of exceptions including but not limited to assert exceptions, are you sure about the logic here?

Reply:
pretty much: if a non-`DecodeError` exception is thrown, something serious went belly-up. usually that implies that the database or the input file is corrupted, or we stepped on a landmine (i.e. an implementation bug). If some data simply cannot be decoded, this should never happen (if it does, I'll consider it to be a bug).

Reply:
found one https://github.com/mercedes-benz/odxtools/blob/8ddf730985d7f92f8399b20a550197ed0783ead3/odxtools/parameters/matchingrequestparameter.py#L36

Reply:
fixed: 62e55eb5d9fea0714d0e9de9cb9cca45717b524d